File: doc/codes/python/examples/guide/custom-pattern.py
import math

import numpy as np

import autd3_pattern as pattern
from autd3.commands import Pattern
from autd3.geometry import Autd3, Geometry
from autd3.units import m, rad, s
from autd3.value import Emission, Intensity, Phase


def main() -> None:
    geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    target = geometry.center() + np.array([0.0, 0.0, 150.0])
    wavelength = pattern.wavelength(340 * m / s)

    # ANCHOR: api
    emissions = geometry.pattern_buffer()
    for slot, device in zip(emissions, geometry):
        for t, pos in enumerate(device.positions()):
            dist = float(np.linalg.norm(target - pos))
            slot[t] = Emission(
                Phase(-dist / wavelength * 2.0 * math.pi * rad),
                Intensity.MAX,
            )

    Pattern(emissions)
    # ANCHOR_END: api


if __name__ == "__main__":
    main()

File: doc/codes/rust/examples/guide/custom-pattern.rs
use core::f32::consts::PI;

use autd3_rs::commands::Pattern;
use autd3_rs::geometry::{Autd3, Geometry, offset};
use autd3_rs::units::{m, mm, rad, s};
use autd3_rs::value::Phase;
use autd3_rs_pattern::wavelength;

fn main() {
    let geometry = Geometry::new(vec![Autd3::default()]);

    let target = geometry.center() + offset(0.0 * mm, 0.0 * mm, 150.0 * mm);
    let wavelength = wavelength(340.0 * m / s);

    // ANCHOR: api
    let mut phases = geometry.phase_buffer();
    for (slot, device) in phases.iter_mut().zip(&geometry) {
        for (p, &pos) in slot.iter_mut().zip(device.positions()) {
            let dist = (target - pos).norm();
            *p = Phase::from(-dist / wavelength.mm() * 2.0 * PI * rad);
        }
    }
    let intensities = geometry.intensity_buffer();

    Pattern::new(&phases, &intensities);
    // ANCHOR_END: api
}
